src/server/game/Entities/Player/Player.cpp

```diff
@@ -16354,22 +16360,48 @@ void Player::SkipQuests(std::vector<uint32> const& questIds)
         UpdateObjectVisibility();
 }
 
-void Player::DespawnPersonalSummonsForQuest(uint32 questId)
+void Player::AddPersonalSummon(ObjectGuid guid)
 {
-    std::list<Creature*> creatureList;
-    GetCreatureListWithOptionsInGrid(creatureList, 100.0f, { .IgnorePhases = true, .PrivateObjectOwnerGuid = GetGUID() }); // we might want to replace this with SummonList in Player at some point
+    _personalSummonGUIDs.push_back(guid);
+}
 
-    for (Creature* creature : creatureList)
+void Player::DespawnAllPersonalSummons()
+{
+    for (ObjectGuid const& guid : _personalSummonGUIDs)
     {
+        if (Creature* creature = ObjectAccessor::GetCreature(*this, guid))
+            creature->DespawnOrUnsummon();
+    }
+    _personalSummonGUIDs.clear();
+}
+
+void Player::DespawnPersonalSummonsForQuest(uint32 questId)
+{
+    GuidVector personalSummonGUIDsCopy = _personalSummonGUIDs;
```

Comment:
In fact, I would entirely replace this for loop and use std::remove_if for cleaner logic:

```
_personalSummonGUIDs.erase(std::remove_if(_personalSummonGUIDs.begin(), _personalSummonGUIDs.end(),
    [&](const ObjectGuid& guid)
    {
        Creature* creature = ObjectAccessor::GetCreature(*this, guid);
        if (!creature)
            return false;

        CreatureSummonedData const* summonedData = sObjectMgr->GetCreatureSummonedData(creature->GetEntry());
        if (!summonedData || !summonedData->DespawnOnQuestsRemoved)
            return false;

        return Trinity::Containers::Contains(*summonedData->DespawnOnQuestsRemoved, questId);
    }),
    _personalSummonGUIDs.end());
```